pandas/_libs/pd_parser.c

```diff
@@ -0,0 +1,178 @@
+
+Distributed under the terms of the BSD Simplified License.
+
+*/
+#define _PANDAS_PARSER_IMPL
+
+#include "pd_parser.h"
+#include "src/parser/io.h"
+
+static int to_double(char *item, double *p_value, char sci, char decimal,
+                            int *maybe_int) {
+  char *p_end = NULL;
+  int error = 0;
+
+  /* Switch to precise xstrtod GH 31364 */
+  *p_value =
+      precise_xstrtod(item, &p_end, decimal, sci, '\0', 1, &error, maybe_int);
+
+  return (error == 0) && (!*p_end);
+}
+
+static int floatify(PyObject *str, double *result, int *maybe_int) {
+  int status;
+  char *data;
+  PyObject *tmp = NULL;
+  const char sci = 'E';
+  const char dec = '.';
+
+  if (PyBytes_Check(str)) {
+    data = PyBytes_AS_STRING(str);
+  } else if (PyUnicode_Check(str)) {
+    tmp = PyUnicode_AsUTF8String(str);
+    if (tmp == NULL) {
+      return -1;
+    }
+    data = PyBytes_AS_STRING(tmp);
+  } else {
+    PyErr_SetString(PyExc_TypeError, "Invalid object type");
+    return -1;
+  }
+
+  status = to_double(data, result, sci, dec, maybe_int);
+
+  if (!status) {
+    /* handle inf/-inf infinity/-infinity */
+    if (strlen(data) == 3) {
+      if (0 == strcasecmp(data, "inf")) {
+        *result = HUGE_VAL;
+        *maybe_int = 0;
+      } else {
+        goto parsingerror;
+      }
+    } else if (strlen(data) == 4) {
+      if (0 == strcasecmp(data, "-inf")) {
+        *result = -HUGE_VAL;
+        *maybe_int = 0;
+      } else if (0 == strcasecmp(data, "+inf")) {
+        *result = HUGE_VAL;
+        *maybe_int = 0;
+      } else {
+        goto parsingerror;
+      }
+    } else if (strlen(data) == 8) {
+      if (0 == strcasecmp(data, "infinity")) {
+        *result = HUGE_VAL;
+        *maybe_int = 0;
+      } else {
+        goto parsingerror;
+      }
+    } else if (strlen(data) == 9) {
+      if (0 == strcasecmp(data, "-infinity")) {
+        *result = -HUGE_VAL;
+        *maybe_int = 0;
+      } else if (0 == strcasecmp(data, "+infinity")) {
+        *result = HUGE_VAL;
+        *maybe_int = 0;
+      } else {
+        goto parsingerror;
+      }
+    } else {
+      goto parsingerror;
+    }
+  }
+
+  Py_XDECREF(tmp);
+  return 0;
+
+parsingerror:
+  PyErr_Format(PyExc_ValueError, "Unable to parse string \"%s\"", data);
+  Py_XDECREF(tmp);
+  return -1;
+}
+
+
+static void pandas_parser_destructor(PyObject *op) {
+  void *ptr = PyCapsule_GetPointer(op, PandasParser_CAPSULE_NAME);
+  PyMem_Free(ptr);
+}
+
+static int pandas_parser_exec(PyObject *module) {
+  PandasParser_CAPI *capi = PyMem_Malloc(sizeof(PandasParser_CAPI));
+  if (capi == NULL) {
+    PyErr_NoMemory();
+    return -1;
+  }
+
+  capi->to_double = to_double;
+  capi->floatify = floatify;
+  capi->new_rd_source = new_rd_source;
+  capi->del_rd_source = del_rd_source;
+  capi->buffer_rd_bytes = buffer_rd_bytes;
+  capi->uint_state_init = uint_state_init;
+  capi->uint64_conflict = uint64_conflict;
+  capi->coliter_setup = coliter_setup;
+  capi->parser_new = parser_new;
```

Comment:
Is it possible to separate actual parsing stuff from string to numeric parsing more?

Reply:
Yea that's a great question. The vast majority of these are used in parsers.pyx; we could implement different capsules for different purposes. It probably would be paired well with a refactor of that Cython module. 

It is hard to tell from the history of parsers.pyx what was placed there as a matter of code organization versus tricks to prevent the build system from getting too heavy handed